lib/jekyll/commands/serve/servlet.rb

```diff
@@ -29,31 +29,38 @@ def search_file(req, res, basename)
         def search_index_file(req, res)
           # /file/index.html -> /file.html
 
-          # First, let's see if the default implementation can figure it out.
-          # (Check for index files in the res.filename directory)
-          if file = super
-            return file
+          # First, let's see if the our super method can figure it out.
+          # (i.e Check for index.html files in the res.filename directory)
+          file = super
+
+          unless file
+            # Ok, I guess that didn't work, I guess there's no basename/index.html
+            # Let's look for basename.html instead...
+
+            # Keep a backup of res.filename in case we need to revert our changes to it
+            old_filename = res.filename
+
+            # We need to calculate the basename and remove it from the path (res.filename)
+            # so we'll turn res.filename into an array of path elements then pop off the
+            # basename.
+            #
+            # We use a while loop just in case res.filename has trailing slashes.
+            #
+            # Once we have popped off the basename, we can join up what's left and use it
+            # as the new res.filename.
+            path_arr = res.filename.scan(%r!/[^/]*!)
+            while basename = path_arr.pop
+              break unless basename == "/"
+            end
```

Comment:
Sorry, didn't see your comment before replying. 3-5? Not sure.

Edge case may not be the best choice of words.

Basically our code doesn't know if `res.filename` will have a trailing `/` or not. We could go with whether it does or doesn't right now, but it may change in the future or (without analysing the WEBrick code) be inconsistent.

Essentially I'm treating it as unsanitised input.